lib/OpenQA/Scheduler/Model/Jobs.pm

```diff
@@ -37,6 +37,82 @@ sub determine_scheduled_jobs ($self) {
     return $self->scheduled_jobs;
 }
 
+# reduces the matching workers of the $to_be_scheduled jobs to support pinning parallel jobs to single host
+sub _pick_slots_with_common_worker_host ($to_be_scheduled) {
+    # return early if we don't need to care about picking a common host for the given set of jobs
+    return undef unless any { $_->{parallel_pinning} } @$to_be_scheduled;
+    my $cluster_size = scalar @$to_be_scheduled;
+    return undef if $cluster_size < 2;
+
+    # pick slots on the first best worker host that has enough matching slots; let each job pick one slot per host
+    my %matching_worker_slots_by_host;
+    my %visited_worker_slots_by_id;
+    my $picked_matching_worker_slots = [];
+    for my $job (@$to_be_scheduled) {
+        # go though the list of matching worker slots and pick one slot per host
+        my %visited_worker_slots_by_host;
+        for my $matching_worker (@{$job->{matching_workers}}) {
+            my $id = $matching_worker->id;
+            next if exists $visited_worker_slots_by_id{$id};  # skip slots that have already been picked
+            my $host = $matching_worker->host;
+            next if $visited_worker_slots_by_host{$host}++;  # skip to pick only one slot per host
+            my $matching_worker_slots = ($matching_worker_slots_by_host{$host} //= []);
+            push @$matching_worker_slots, $matching_worker;
+            $visited_worker_slots_by_id{$id} = $job;
+            next if @$matching_worker_slots < $cluster_size;
+            $picked_matching_worker_slots = $matching_worker_slots;
+            last;
+        }
+
+        # go tough the list of matching workers again to re-visit hosts we could not pick a slot on
+        for my $matching_worker (@{$job->{matching_workers}}) {
+            my $host = $matching_worker->host;
+            next if $visited_worker_slots_by_host{$host};  # skip hosts we were able to pick a slot on
+
+            # check the job we are competing with for this slot and see whether we might be able to swap picks by finding
```

Comment:
I assume this is only when the candidate host already has other jobs planned occupying jobs? In such case I would be ok if the pinned cluster simply doesn't get a chance to run until enough free slots are available. Can this help to keep the code simpler?

Reply:
> In such case I would be ok if the pinned cluster simply doesn't get a chance to run until enough free slots are available.

But it might never run at all then. I think it would be nice if one has a setup with a limited number of slots and parallelism at the same time when assignments would work at all (at least in most cases).